### PR TITLE
aya: fix proc maps parsing for deleted files

### LINKED ISSUE #1423: uprobe: find_library_path_by_name fails when /proc/pid/maps contains "(deleted)" suffix
### Description
On certain environments (like Android or systems where libraries are updated/deleted), `/proc/pid/maps` entries may contain a ` (deleted)` suffix. For example:
`7f1bca83a000-7f1bca83c000 rw-p 00036000 fd:01 2895508 /usr/lib/libc.so (deleted)`

The current `ProcMapEntry::parse` logic is too strict and expects the path to be the very last token. This causes the entire parsing process to fail, preventing uprobes from being attached.

### Version
Aya: 0780eef64
Kernel: 5.10
android: 16

### PR DESCRIPTION
**Pull request by Satar07**
See commit message.

Fixes #1423.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aya-rs/aya/1424)

### CODE REVIEW

**Comment by Satar07** [comment]
Thanks! I've updated the logic to strictly accept only the (deleted) suffix after an absolute path. Any other trailing tokens will still cause a parse error. I also added a test case to cover this.

File: aya/src/programs/uprobe.rs [hunk 1]
```diff
@@ -398,24 +398,35 @@ impl<'a> ProcMapEntry<'a> {
...
-                        Some(Ok(path))
+        let path_part = parts.next();
+        let path = match path_part {
+            Some(val) => {
+                // Special mappings (e.g., [stack], [vdso], or Android's [anon:...]).
+                if val.first() == Some(&b'[') {
+                    None
+                }
+                // Absolute paths.
+                else if val.first() == Some(&b'/') {
```
Comment (vadorovsky) [inline aya/src/programs/uprobe.rs]: nit: Assign the result of `val.first()` to a variable, so it does not end up being called twice:

```suggestion
                let first = val.first();
                if first == Some(&b'[') {
                    None
                }
                // Absolute paths.
                else if first == Some(&b'/') {
```

**Comment by tamird** [comment]
As mentioned in an earlier comment that was perhaps missed: can we do this with less rewriting in the LLM's preferred style?

**Comment by Satar07** [comment]
Apologies for the noise. To be honest, I ran into this issue while working on an Android eBPF tool and used an LLM to help quickly prototype a fix. I was eager to contribute the solution back upstream but realized I shouldn't have pushed the generated code without refining it first.

I genuinely appreciate the rigorous review process here—it clearly shows why Aya is such a high-quality library.

I'll rework the implementation to address this.

**Comment by tamird** [comment]
@Satar07 I pushed another commit to reduce the diff further. Does this look reasonable to you?

**Comment by Satar07** [comment]
That works for me! I appreciate you jumping in to clean this up.

**Comment by Satar07** [comment]
@tamird  Merry Christmas! 🎄

I wanted to give you an update—I tested the latest version on Android, but unfortunately, it still fails on several common patterns found there. Specifically, I encountered these cases:

```
/memfd:jit-zygote-cache (deleted)
[anon:dalvik-Boot image reservation]
[anon:dalvik-zygote space]
[anon:dalvik-Sentinel fault page]
[page size compat]
[anon:cfi shadow]
/dev/ashmem/CursorWindow: /data/user/0/package/databases/kitefly.db (deleted)
```

I’ve written a new version that handles these scenarios and verified that it works in practice. I also added tests to cover these new edge cases.

Could you take a look at the rewritten path parsing logic?

One specific detail involves how I handled `/dev/ashmem`. I'd like to discuss if this approach is reasonable:

```Rust
let path = match body {
    [] if !is_deleted => Ok(None),
    [first, ..]
        if first.starts_with(b"[")
            && body.last().unwrap().ends_with(b"]")
            && !is_deleted =>
    {
        Ok(None)
    }
    // Specific handling for ashmem
    [first, ..] if first.starts_with(b"/dev/ashmem") => Ok(None),
    [bytes] => {
        let path = Path::new(OsStr::from_bytes(bytes));
        if path.is_absolute() {
            Ok(Some(path))
        } else {
            Err(err())
        }
    }
    _ => Err(err()),
}?;
```

I singled out `/dev/ashmem` because it's arguably the weirdest case—it starts like a valid path but often has spaces or extra segments. I decided to return `Ok(None)` to simply ignore it. Do you think that's the right call? regardless, since these lines appear in actual Android maps, they definitely shouldn't trigger a parse error.

**Comment by tamird** [comment]
@Satar07 I noticed your new tests have two spaces before e.g. `[page size compat]`. Is that intentional?

**Comment by Satar07** [comment]
@tamird Yes, that was intentional. I noticed that on Android, the proc/maps output often uses a large number of spaces (rather than tabs) to align the pathname column.

```
1|emu64a:/ # cat /proc/11306/maps | grep '\[' | head -n 10 | od -tx1 -tc
0000000  31  34  30  30  30  30  30  30  2d  33  34  30  30  30  30  30
          1   4   0   0   0   0   0   0   -   3   4   0   0   0   0   0
0000020  30  20  72  77  2d  70  20  30  30  30  30  30  30  30  30  20
          0       r   w   -   p       0   0   0   0   0   0   0   0
0000040  30  30  3a  30  30  20  30  20  20  20  20  20  20  20  20  20
          0   0   :   0   0       0
0000060  20  20  20  20  20  20  20  20  20  20  20  20  20  20  20  20

0000100  20  20  20  20  20  20  20  20  20  5b  61  6e  6f  6e  3a  64
                                              [   a   n   o   n   :   d
0000120  61  6c  76  69  6b  2d  6d  61  69  6e  20  73  70  61  63  65
          a   l   v   i   k   -   m   a   i   n       s   p   a   c   e
0000140  5d  0a  33  34  30  30  30  30  30  30  2d  35  34  30  30  30
          ]  \n   3   4   0   0   0   0   0   0   -   5   4   0   0   0
```

**Comment by tamird** [comment]
Cool, thanks for clarifying. I had to force push to pick up #1431 so I could test locally. How does this look to you?

**Comment by Satar07** [comment]
Looks good to me. Feel free to do whatever works best for you.

**Review by tamird — APPROVED** [review]
@codex review

@vadorovsky could you have another look please?

> @tamird reviewed 6 files and all commit messages, made 1 comment, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1424#-:-Ohu2n7IDT43DXU3FQEL:b829r8f)* status: all files reviewed, 3 unresolved discussions (waiting on @Satar07).

File: aya/src/programs/uprobe.rs [hunk 2]
```diff
@@ -337,13 +340,19 @@ pub enum ProcMapError {
...
-        line: Vec<u8>,
+        line: OsString,
     },
 }
 
+#[derive(Debug)]
+struct ProcMapPath<'a> {
+    path: &'a Path,
+    deleted: bool,
+}
```
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/programs/uprobe.rs]: Removed.

File: aya/src/programs/uprobe.rs [hunk 3]
```diff
@@ -362,19 +371,44 @@ struct ProcMapEntry<'a> {
     dev: &'a OsStr,
     #[cfg_attr(not(test), expect(dead_code))]
     inode: u32,
-    path: Option<&'a Path>,
+    path: Option<ProcMapPath<'a>>,
+}
+
```
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/programs/uprobe.rs:328]: Done.

File: aya/src/programs/uprobe.rs [hunk 4]
```diff
@@ -422,11 +456,27 @@ impl<'a> ProcMapEntry<'a> {
...
                 [b'[', .., b']'] => None,
                 path => {
-                    let path = Path::new(OsStr::from_bytes(path));
-                    if !path.is_absolute() {
-                        Some(Err(err()))
+                    let path = path
+                        .split(u8::is_ascii_whitespace)
+                        .filter(|part| !part.is_empty())
+                        .collect::<Vec<_>>();
+                    let path = match path.as_slice() {
```
[the remark this answers is not in the document]
Comment (tamird) [reply aya/src/programs/uprobe.rs]: Removed.

**Comment by tamird** [comment]
OK I've made another big update here: we no longer try to do anything fancy with "path" - we don't even assume it's a path. We leave it to the caller to deal with it, which feels like the right API boundary. I'll wait for @vadorovsky to have a look before merging. @Satar07 please also have a look.
